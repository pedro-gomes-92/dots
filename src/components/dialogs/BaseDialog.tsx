import React, { FunctionComponent, useState } from 'react';
import classnames from 'classnames';

import { BaseProps, Base } from '../Base';
import { OverlayProps, Overlay, Close } from '../miscellaneous';
import { CenterLayout } from '../layouts';

export interface BaseDialogProps extends BaseProps {
  onClose?: OverlayProps['onClick'];
  onOutsideClick?: OverlayProps['onClick'];
}

export const BaseDialog: FunctionComponent<BaseDialogProps> = ({
  className,
  children,
  onClose,
  onOutsideClick,
  isVisible,
  ...rest
}: BaseDialogProps): JSX.Element => {
  const [visible, setVisible] = useState(isVisible);

  let handleClose;
  if (onClose) {
    handleClose = (event: React.MouseEvent<HTMLElement, MouseEvent>): void => {
      setVisible(false);
      onClose(event);
    };
  }

  let handleOutsideClick = handleClose;
  if (onOutsideClick) {
    handleOutsideClick = (event: React.MouseEvent<HTMLElement, MouseEvent>): void => {
      setVisible(false);
      onOutsideClick(event);
    };
  }

  return (
    <Base className={classnames('dialog', className)} isVisible={visible} {...rest}>
      <CenterLayout>
        <Overlay onClick={handleOutsideClick} />
        {children}
      </CenterLayout>
      {onClose && <Close size="large" onClick={handleClose} />}
    </Base>
  );
};

BaseDialog.defaultProps = {
  onClose: (): void => {},
  isVisible: true,
};
