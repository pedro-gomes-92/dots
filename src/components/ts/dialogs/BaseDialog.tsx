import React, { FunctionComponent, useState } from 'react';
import classnames from 'classnames';

import { BaseProps, Base } from '../Base';
import { OverlayProps, Overlay } from '../miscellaneous';
import { CenterLayout } from '../layouts';
import { CloseDialog } from './CloseDialog';

export interface BaseDialogProps extends BaseProps {
  onClose?: OverlayProps['onClick'];
  onOutsideClick?: OverlayProps['onClick'];
}

export const BaseDialog: FunctionComponent<BaseDialogProps> = ({
  className,
  children,
  onClose,
  onOutsideClick,
  isVisible,
  animations,
  ...rest
}: BaseDialogProps): JSX.Element => {
  const [visible, setVisible] = useState(isVisible);

  const [showAnimation, hideAnimation] = animations;

  let handleClose;
  if (onClose) {
    handleClose = (event: React.MouseEvent<HTMLElement, MouseEvent>): void => {
      setVisible(false);
      onClose(event);
    };
  }

  let handleOutsideClick = handleClose;
  if (onOutsideClick) {
    handleOutsideClick = (event: React.MouseEvent<HTMLElement, MouseEvent>): void => {
      setVisible(false);
      onOutsideClick(event);
    };
  }

  return (
    <Base
      className={classnames('dialog', className)}
      animations={visible ? [showAnimation] : [hideAnimation]}
      {...rest}
    >
      <CenterLayout>
        <Overlay onClick={handleOutsideClick} />
        {children}
      </CenterLayout>
      {onClose && <CloseDialog size="large" onClick={handleClose} />}
    </Base>
  );
};

BaseDialog.defaultProps = {
  onClose: (): void => {},
  isVisible: true,
  animations: [{ name: 'fadeIn', duration: 'fast' }, { name: 'fadeOut', duration: 'fast' }],
};
