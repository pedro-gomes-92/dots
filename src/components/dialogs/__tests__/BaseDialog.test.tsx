import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { BaseDialog } from '../BaseDialog';

describe('Given <BaseDialog />', (): void => {
  let container: HTMLElement;
  describe('When initializes', (): void => {
    it('Then should match the snapshot', (): void => {
      container = render(<BaseDialog />).container;
      expect(container).toMatchSnapshot();
    });

    describe('And uses onClose', (): void => {
      let handleClose: jest.Mock;
      let close: HTMLElement;

      beforeEach((): void => {
        handleClose = jest.fn();
        container = render(<BaseDialog onClose={handleClose} />).container;
        close = container.querySelector('.close');

        fireEvent.click(close);
      });

      it('Then should match the snapshot', (): void => {
        expect(container).toMatchSnapshot();
      });

      it('And should render close button', (): void => {
        expect(close).toBeTruthy();
      });

      it('And should hide the dialog', (): void => {
        expect(container.querySelector('.dialog').classList).toContainEqual('is-hidden');
      });

      it('And should trigger onClose', (): void => {
        expect(handleClose).toHaveBeenCalledTimes(1);
      });
    });

    describe('And uses onOutsideClick', (): void => {
      let handleOutsideClick: jest.Mock;
      let overlay: HTMLElement;

      beforeEach((): void => {
        handleOutsideClick = jest.fn();
        container = render(<BaseDialog onOutsideClick={handleOutsideClick} />).container;
        overlay = container.querySelector('.overlay');

        fireEvent.click(overlay);
      });

      it('Then should hide the dialog', (): void => {
        expect(container.querySelector('.dialog').classList).toContainEqual('is-hidden');
      });

      it('And should trigger onOutsideClick', (): void => {
        expect(handleOutsideClick).toHaveBeenCalledTimes(1);
      });
    });
  });
});
