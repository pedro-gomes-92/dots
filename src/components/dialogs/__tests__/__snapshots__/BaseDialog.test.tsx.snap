// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given <BaseDialog /> When initializes And uses onClose Then should match the snapshot 1`] = `
<div>
  <div
    class="is-hidden dialog"
  >
    <div
      class="layout is-0 layout-center"
    >
      <div
        class="layout-item"
      >
        <div
          class="overlay is-clickable"
        />
      </div>
    </div>
    <button
      class="is-large button is-text close"
    >
      <span
        class="icon"
      >
        <i
          class="is-large icon-font icon-font-close"
        />
      </span>
    </button>
  </div>
</div>
`;

exports[`Given <BaseDialog /> When initializes Then should match the snapshot 1`] = `
<div>
  <div
    class="dialog"
  >
    <div
      class="layout is-0 layout-center"
    >
      <div
        class="layout-item"
      >
        <div
          class="overlay is-clickable"
        />
      </div>
    </div>
    <button
      class="is-large button is-text close"
    >
      <span
        class="icon"
      >
        <i
          class="is-large icon-font icon-font-close"
        />
      </span>
    </button>
  </div>
</div>
`;
